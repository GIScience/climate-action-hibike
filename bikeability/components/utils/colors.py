from typing import Union

import matplotlib
import pandas as pd
from matplotlib.colors import Normalize, to_hex
from pydantic_extra_types.color import Color

from bikeability.components.dooring_risk.dooring_risk import DooringRiskCategory
from bikeability.components.path_categories.path_categories import PathCategory
from bikeability.components.smoothness.smoothness import SmoothnessCategory
from bikeability.components.surface_types.surface_types import SurfaceType


def get_qualitative_color(
    category: Union[PathCategory, SmoothnessCategory, SurfaceType, DooringRiskCategory], cmap_name: str
) -> Color:
    norm = Normalize(0, 1)
    cmap = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap_name).get_cmap()
    cmap.set_under('#808080')

    category_norm = {name: idx / (len(category.get_visible()) - 1) for idx, name in enumerate(category.get_visible())}

    if category.value == 'unknown':
        return Color(to_hex(cmap(-9999)))
    else:
        return Color(to_hex(cmap(category_norm[category])))


def get_continuous_colors(category: pd.Series, cmap_name: str) -> list[Color]:
    norm = matplotlib.colors.Normalize(vmin=0, vmax=1)
    cmap = matplotlib.colormaps.get(cmap_name)
    cmap.set_bad('#808080')
    mapped_colors = [Color(matplotlib.colors.to_hex(col)) for col in cmap(norm(category))]
    return mapped_colors
