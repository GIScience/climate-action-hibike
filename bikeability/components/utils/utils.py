import logging
from enum import StrEnum

import geopandas as gpd
import shapely
from climatoology.utility.exception import ClimatoologyUserError
from ohsome import OhsomeClient
from pyproj import CRS, Transformer
from shapely.ops import transform

log = logging.getLogger(__name__)


class Topics(StrEnum):
    TRAFFIC = 'traffic'
    SURFACE = 'surface'
    SUMMARY = 'summary'
    CONNECTIVITY = 'connectivity'
    BARRIERS = 'barriers'
    SAFETY = 'safety'
    GREENNESS = 'greenness'


def check_paths_count_limit(aoi: shapely.MultiPolygon, ohsome: OhsomeClient, count_limit: int) -> None:
    """
    Check whether paths count is over than limit. (NOTE: just check path_lines)
    """

    ohsome_responses = ohsome.elements.count.post(bpolys=aoi, filter=ohsome_filter('line')).data
    path_lines_count = sum([response['value'] for response in ohsome_responses['result']])
    log.info(f'There are {path_lines_count} are selected.')
    if path_lines_count > count_limit:
        raise ClimatoologyUserError(
            f'There are too many path segments in the selected area: {path_lines_count} path segments. '
            f'Currently, only areas with a maximum of 500,000 path segments are allowed. '
            f'Please select a smaller area or a sub-region of your selected area.'
        )


def fetch_osm_data(aoi: shapely.MultiPolygon, osm_filter: str, ohsome: OhsomeClient) -> gpd.GeoDataFrame:
    elements = ohsome.elements.geometry.post(
        bpolys=aoi, clipGeometry=True, properties='tags', filter=osm_filter
    ).as_dataframe()
    elements = elements.reset_index()
    return elements[['@osmId', 'geometry', '@other_tags']]


def ohsome_filter(geometry_type: str) -> str:
    return str(
        f'geometry:{geometry_type} and '
        '(highway=* or railway=platform) and not '
        '(cycleway=separate or cycleway:both=separate or '
        '(cycleway:right=separate and cycleway:left=separate) or '
        'indoor=yes or indoor=corridor)'
    )


def get_utm_zone(aoi: shapely.MultiPolygon) -> CRS:
    return gpd.GeoSeries(data=aoi, crs='EPSG:4326').estimate_utm_crs()


def get_buffered_aoi(aoi: shapely.MultiPolygon) -> shapely.MultiPolygon:
    wgs84 = CRS('EPSG:4326')
    utm = get_utm_zone(aoi)

    geographic_projection_function = Transformer.from_crs(wgs84, utm, always_xy=True).transform
    wgs84_projection_function = Transformer.from_crs(utm, wgs84, always_xy=True).transform
    projected_aoi = transform(geographic_projection_function, aoi)
    # changed the distance to a fixed value of 5 km.
    buffered_aoi = projected_aoi.buffer(5000)
    return transform(wgs84_projection_function, buffered_aoi)


def calculate_length(length_resolution_m, paths, projected_crs):
    stats = paths.copy()
    stats = stats.loc[stats.geometry.geom_type.isin(('MultiLineString', 'LineString'))]
    stats = stats.to_crs(projected_crs)
    stats['length'] = stats.length / length_resolution_m
    stats['length'] = round(stats['length'], 2)
    return stats
