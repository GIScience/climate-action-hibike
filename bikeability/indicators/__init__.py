from .path_categories import *
from .smoothness import *
from .surface_types import *
