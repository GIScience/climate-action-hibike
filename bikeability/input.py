from typing import Optional, Self, Dict

from pydantic import BaseModel, Field, model_validator

from bikeability.indicators.path_categories import PathCategory
from bikeability.indicators.smoothness import SmoothnessCategory
from bikeability.indicators.dooring_risk import DooringRiskCategory


class PathRating(BaseModel):
    designated_exclusive: float = Field(
        title='Designated Exclusive Path Rating',
        description='Qualitative rating (between 0..1) of paths that are designated for cycling and not shared with pedestrians.',
        ge=0,
        le=1,
        examples=[1.0],
        default=1.0,
    )

    designated_shared_with_pedestrians: float = Field(
        title='Designated Shared with Pedestrians Path Rating',
        description='Qualitative rating (between 0..1) of paths that are designated for cycling and shared with pedestrians.',
        ge=0,
        le=1,
        examples=[0.8],
        default=0.8,
    )

    shared_with_motorised_traffic_walking_speed: float = Field(
        title='Shared with Motorised Traffic Walking Speed Path Rating',
        description='Qualitative rating (between 0..1) of paths that are shared with motorised traffic with speed limits of roughly walking speed (<=15 km/h).',
        ge=0,
        le=1,
        examples=[0.8],
        default=0.8,
    )

    shared_with_motorised_traffic_low_speed: float = Field(
        title='Shared with Motorised Traffic Low Speed Path Rating',
        description='Qualitative rating (between 0..1) of paths that are shared with motorised traffic with max speed limits of 30 km/h.',
        ge=0,
        le=1,
        examples=[0.6],
        default=0.6,
    )

    shared_with_motorised_traffic_medium_speed: float = Field(
        title='Shared with Motorised Traffic Medium Speed Path Rating',
        description='Qualitative rating (between 0..1) of paths that are shared with motorised traffic with max speed limits of 50 km/h.',
        ge=0,
        le=1,
        examples=[0.4],
        default=0.4,
    )

    shared_with_motorised_traffic_high_speed: float = Field(
        title='Shared with Motorised Traffic High Speed Path Rating',
        description='Qualitative rating (between 0..1) of paths that are shared with motorised traffic with max speed limits of 100 km/h.',
        ge=0,
        le=1,
        examples=[0.1],
        default=0.1,
    )

    requires_dismounting: float = Field(
        title='Requires Dismounting Path Rating',
        description='Qualitative rating (between 0..1) of paths where cyclists must dismount.',
        ge=0,
        le=1,
        examples=[0.05],
        default=0.05,
    )

    not_bikeable: float = Field(
        title='Not Bikeable Path Rating',
        description='Qualitative rating (between 0..1) of paths that are not bikeable.',
        ge=0,
        le=1,
        examples=[0.0],
        default=0.0,
    )

    unknown: float = Field(
        title='Unknown Path Rating',
        description='Qualitative (between 0..1) rating of paths that are in principle bikeable but are could not be categorised (default -9999, which is out of scale)',
        ge=0,
        le=1,
        examples=[0.0],
        default=-9999,
    )

    pedestrian_exclusive: float = Field(
        title='Pedestrian Exclusive Path Rating',
        description='Shadow category (not used in Bikeable Path Category Indicator)',
        ge=0,
        le=1,
        examples=[0.0],
        default=-9999,
    )

    restricted_access: float = Field(
        title='Restricted Access Path Rating',
        description='Shadow category (not used in Bikeable Path Category Indicator)',
        ge=0,
        le=1,
        examples=[0.0],
        default=-9999,
    )

    @model_validator(mode='after')
    def check_order(self) -> Self:
        assert (
            0
            <= self.not_bikeable
            <= self.requires_dismounting
            <= self.shared_with_motorised_traffic_high_speed
            <= self.shared_with_motorised_traffic_medium_speed
            <= self.shared_with_motorised_traffic_low_speed
            <= self.shared_with_motorised_traffic_walking_speed
            <= self.designated_shared_with_pedestrians
            <= self.designated_exclusive
            <= 1
        ), 'Qualitative rating must respect semantic order of categories!'
        return self


class PathSmoothnessRating(BaseModel):
    excellent: float = Field(
        title='Excellent Smoothness Rating',
        description='Qualitative rating (between 0..1) of paths with excellent smoothness.',
        ge=0,
        le=1,
        examples=[1.0],
        default=1.0,
    )

    good: float = Field(
        title='Good Smoothness Rating',
        description='Qualitative rating (between 0..1) of paths with good smoothness.',
        ge=0,
        le=1,
        examples=[0.75],
        default=0.75,
    )

    intermediate: float = Field(
        title='Intermediate Smoothness Rating',
        description='Qualitative rating (between 0..1) of paths with intermediate smoothness.',
        ge=0,
        le=1,
        examples=[0.5],
        default=0.5,
    )

    bad: float = Field(
        title='Bad Smoothness Rating',
        description='Qualitative rating (between 0..1) of paths with bad smoothness.',
        ge=0,
        le=1,
        examples=[0.25],
        default=0.25,
    )

    too_bumpy_to_ride: float = Field(
        title='Too Bumpy to Ride Smoothness Rating',
        description='Qualitative rating (between 0..1) of paths with very bad or worse smoothness.',
        ge=0,
        le=1,
        examples=[0.0],
        default=0.0,
    )

    unknown: float = Field(
        title='Unknown Smoothness Rating',
        description='Qualitative (between 0..1) rating of paths that could not be categorised (default -9999, which is out of scale)',
        ge=0,
        le=1,
        examples=[0.0],
        default=-9999,
    )

    @model_validator(mode='after')
    def check_order(self) -> Self:
        assert (
            0 <= self.too_bumpy_to_ride <= self.bad <= self.intermediate <= self.good <= self.excellent <= 1
        ), 'Qualitative rating must respect semantic order of categories!'
        return self


class PathDooringRiskRating(BaseModel):
    safe_route: float = Field(
        title='Safe Route Dooring Risk Rating',
        description='Qualitative rating (between 0..1) of paths with no or significantly less risk of dooring.',
        ge=0,
        le=1,
        examples=[1.0],
        default=1.0,
    )

    risk_of_dooring: float = Field(
        title='Risk of Dooring Risk Rating',
        description='Qualitative rating (between 0..1) of paths with a high risk of dooring.',
        ge=0,
        le=1,
        examples=[0.0],
        default=0.0,
    )

    unknown: float = Field(
        title='Unknown Dooring Risk Rating',
        description='Qualitative (between 0..1) rating of paths whose dooring risk could not be determined (default -9999, which is out of scale)',
        ge=0,
        le=1,
        examples=[0.0],
        default=-9999,
    )

    @model_validator(mode='after')
    def check_order(self) -> Self:
        assert (
            0 <= self.risk_of_dooring <= self.safe_route <= 1
        ), 'Qualitative rating must respect semantic order of categories!'
        return self


class ComputeInputBikeability(BaseModel):
    path_rating: Optional[PathRating] = Field(
        title='Path Rating Mapping',
        description='Qualitative rating for each of the available path categories.',
        examples=[PathRating()],
        default=PathRating(),
    )

    smoothness_rating: Optional[PathSmoothnessRating] = Field(
        title='Path Smoothness Mapping',
        description='Qualitative rating for each available smoothness category',
        examples=[PathSmoothnessRating()],
        default=PathSmoothnessRating(),
    )

    dooring_risk_rating: Optional[PathDooringRiskRating] = Field(
        title='Path Dooring Risk Rating',
        description='Qualitative rating for dooring risk',
        examples=[PathDooringRiskRating()],
        default=PathDooringRiskRating(),
    )

    def get_path_rating_mapping(self) -> Dict[PathCategory, float]:
        mapping = self.path_rating.model_dump()
        return {PathCategory(k): v for k, v in mapping.items()}

    def get_path_smoothness_mapping(self) -> Dict[SmoothnessCategory, float]:
        mapping = self.smoothness_rating.model_dump()
        return {SmoothnessCategory(k): v for k, v in mapping.items()}

    def get_path_dooring_mapping(self) -> Dict[DooringRiskCategory, float]:
        mapping = self.dooring_risk_rating.model_dump()
        return {DooringRiskCategory(k): v for k, v in mapping.items()}
