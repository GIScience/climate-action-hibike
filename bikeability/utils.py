import logging
from typing import Union

import geopandas as gpd
import matplotlib
import pandas as pd
import shapely
from climatoology.utility.exception import ClimatoologyUserError
from matplotlib.colors import Normalize, to_hex
from ohsome import OhsomeClient
from pydantic_extra_types.color import Color
from pyproj import CRS, Transformer
from shapely.ops import transform

from bikeability.indicators.dooring_risk import DooringRiskCategory
from bikeability.indicators.path_categories import PathCategory
from bikeability.indicators.smoothness import SmoothnessCategory
from bikeability.indicators.surface_types import SurfaceType

log = logging.getLogger(__name__)


def check_paths_count_limit(aoi: shapely.MultiPolygon, ohsome: OhsomeClient, count_limit: int) -> None:
    """
    Check whether paths count is over than limit. (NOTE: just check path_lines)
    """

    ohsome_responses = ohsome.elements.count.post(bpolys=aoi, filter=ohsome_filter('line')).data
    path_lines_count = sum([response['value'] for response in ohsome_responses['result']])
    log.info(f'There are {path_lines_count} are selected.')
    if path_lines_count > count_limit:
        raise ClimatoologyUserError(
            f'There are too many path segments in the selected area: {path_lines_count} path segments. '
            f'Currently, only areas with a maximum of 500,000 path segments are allowed. '
            f'Please select a smaller area or a sub-region of your selected area.'
        )


def fetch_osm_data(aoi: shapely.MultiPolygon, osm_filter: str, ohsome: OhsomeClient) -> gpd.GeoDataFrame:
    elements = ohsome.elements.geometry.post(
        bpolys=aoi, clipGeometry=True, properties='tags', filter=osm_filter
    ).as_dataframe()
    elements = elements.reset_index()
    return elements[['@osmId', 'geometry', '@other_tags']]


def get_qualitative_color(
    category: Union[PathCategory, SmoothnessCategory, SurfaceType, DooringRiskCategory], cmap_name: str
) -> Color:
    norm = Normalize(0, 1)
    cmap = matplotlib.cm.ScalarMappable(norm=norm, cmap=cmap_name).get_cmap()
    cmap.set_under('#808080')

    category_norm = {name: idx / (len(category.get_visible()) - 1) for idx, name in enumerate(category.get_visible())}

    if category.value == 'unknown':
        return Color(to_hex(cmap(-9999)))
    else:
        return Color(to_hex(cmap(category_norm[category])))


def get_continuous_colors(category: pd.Series, cmap_name: str) -> list[Color]:
    norm = matplotlib.colors.Normalize(vmin=0, vmax=1)
    cmap = matplotlib.colormaps.get(cmap_name)
    cmap.set_bad('#808080')
    mapped_colors = [Color(matplotlib.colors.to_hex(col)) for col in cmap(norm(category))]
    return mapped_colors


def ohsome_filter(geometry_type: str) -> str:
    return str(
        f'geometry:{geometry_type} and '
        '(highway=* or railway=platform) and not '
        '(cycleway=separate or cycleway:both=separate or '
        '(cycleway:right=separate and cycleway:left=separate) or '
        'indoor=yes or indoor=corridor)'
    )


def get_utm_zone(aoi: shapely.MultiPolygon) -> CRS:
    return gpd.GeoSeries(data=aoi, crs='EPSG:4326').estimate_utm_crs()


def get_buffered_aoi(aoi: shapely.MultiPolygon) -> shapely.MultiPolygon:
    wgs84 = CRS('EPSG:4326')
    utm = get_utm_zone(aoi)

    geographic_projection_function = Transformer.from_crs(wgs84, utm, always_xy=True).transform
    wgs84_projection_function = Transformer.from_crs(utm, wgs84, always_xy=True).transform
    projected_aoi = transform(geographic_projection_function, aoi)
    # changed the distance to a fixed value of 5 km.
    buffered_aoi = projected_aoi.buffer(5000)
    return transform(wgs84_projection_function, buffered_aoi)


def zebra_crossings_filter() -> str:
    return str('geometry:point and type:node and (crossing=zebra or crossing:markings=zebra or crossing_ref=zebra)')


def parallel_parking_filter(geometry_type) -> str:
    return str(f'geometry:{geometry_type} and amenity=parking and orientation=parallel')
