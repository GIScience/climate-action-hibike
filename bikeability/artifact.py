from pathlib import Path

import geopandas as gpd
import pandas as pd
import shapely
from climatoology.base.artifact import (
    _Artifact,
    create_geojson_artifact,
)
from climatoology.base.computation import ComputationResources

from bikeability.input import PathRating
from bikeability.utils import (
    pathratings_legend_fix,
    get_qualitative_color,
    PathCategory,
)


def build_paths_artifact(
    paths_line: gpd.GeoDataFrame,
    paths_polygon: gpd.GeoDataFrame,
    ratings: PathRating,
    clip_aoi: shapely.MultiPolygon,
    resources: ComputationResources,
    cmap_name: str = 'RdYlBu_r',
) -> _Artifact:
    paths_line = paths_line.clip(clip_aoi, keep_geom_type=True)
    paths_polygon = paths_polygon.clip(clip_aoi, keep_geom_type=True)
    paths_all = pd.concat([paths_line, paths_polygon], ignore_index=True)

    paths_all['color'] = paths_all.category.apply(get_qualitative_color, cmap_name=cmap_name)
    return create_geojson_artifact(
        features=paths_all.geometry,
        layer_name='Cycling infrastructure path categories',
        caption=Path('resources/info/path_categories/caption.md').read_text(),
        description=Path('resources/info/path_categories/description.md').read_text(),
        label=paths_all.category.apply(lambda r: r.name).to_list(),
        color=paths_all.color.to_list(),
        legend_data={
            pathratings_legend_fix.get(category.value, category.value): get_qualitative_color(category, cmap_name)
            for category in PathCategory
        },
        resources=resources,
        filename='cycling_infrastructure_path_categories',
    )
