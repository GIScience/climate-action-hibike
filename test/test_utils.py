from functools import partial

import geopandas as gpd
import geopandas.testing
import pandas as pd
import pytest
import shapely
from approvaltests.approvals import verify
from approvaltests.namer import NamerFactory
from ohsome import OhsomeClient, OhsomeResponse
from pydantic_extra_types.color import Color
from shapely.testing import assert_geometries_equal
from urllib3 import Retry

from bikeability.utils import (
    PathCategory,
    fetch_osm_data,
    get_color,
    apply_path_category_filters,
    boost_route_members,
    fix_geometry_collection,
    ohsome_filter,
)

validation_objects = {
    PathCategory.NOT_BIKEABLE: {'way/4084008', 'way/24635973', 'way/343029968'},
    # https://www.openstreetmap.org/way/4084008 highway=trunk
    # https://www.openstreetmap.org/way/24635973 highway=secondary and bicycle=no
    # https://www.openstreetmap.org/way/343029968 highway=primary and motorroad=yes
    PathCategory.DESIGNATED_EXCLUSIVE: {'way/246387137', 'way/118975501'},
    # https://www.openstreetmap.org/way/246387137 highway=cycleway and foot=no
    # https://www.openstreetmap.org/way/118975501 highway=path and foot=yes and segregated=yes
    PathCategory.DESIGNATED_SHARED_WITH_PEDESTRIANS: {'way/587937936', 'way/27620739', 'way/406929620'},
    # https://www.openstreetmap.org/way/587937936 highway=path and bicycle=designated and foot=designated and segregated=no
    # https://www.openstreetmap.org/way/27620739 highway=footway and bicycle=yes but no foot tag
    # https://www.openstreetmap.org/way/406929620 highway=track
    PathCategory.SHARED_WITH_MOTORISED_TRAFFIC_WALKING_SPEED: {
        'way/34386123',
        'way/27342468',
        'way/715905252',
        'way/35978590',
        'way/274709010',
        'way/440466305',
        'way/83188869',
        'way/191212309',
    },
    # https://www.openstreetmap.org/way/34386123 highway=living_street and bicycle=yes
    # https://www.openstreetmap.org/way/27342468 highway=living_street cycleway:both=no
    # https://www.openstreetmap.org/way/715905252 highway=service and bicycle=yes
    # https://www.openstreetmap.org/way/35978590 highway=service and no bicycle tag
    # https://www.openstreetmap.org/way/274709010 highway=residential and maxspeed=10 and cycleway:both=no
    # https://www.openstreetmap.org/way/440466305 highway=living_street and maxspeed=walk
    # https://www.openstreetmap.org/way/83188869 highway=residential and maxspeed=15 and cycleway:both=no
    # https://www.openstreetmap.org/way/191212309 highway=residential and bicylce=designated ("Fahrradstraße") and maxspeed=15 and motorvehicle=destination
    PathCategory.SHARED_WITH_MOTORISED_TRAFFIC_LOW_SPEED: {
        'way/37685272',
        'way/244287984',
        'way/258562283',
        'way/28891216',
        'way/932097064',
        'way/277455037',
    },
    # https://www.openstreetmap.org/way/37685272 highway=residential and bicycle=designated and bicycle_road=yes ('Fahrradstraße') and maxspeed=30 and motor_vehicle=yes
    # https://www.openstreetmap.org/way/244287984 highway=residential and cycleway:both=no and maxspeed=20
    # https://www.openstreetmap.org/way/258562283 highway=tertiary and cycleway:left=no and cycleway:right=lane and maxspeed=30
    # https://www.openstreetmap.org/way/28891216 highway=primary cycleway:both=lane and maxspeed=30
    # https://www.openstreetmap.org/way/932097064 highway=residential and maxspeed=30 and cycleway:both=no
    # https://www.openstreetmap.org/way/277455037 highway=tertiary and zone:maxspeed=DE:30 and cycleway:right=no
    PathCategory.SHARED_WITH_MOTORISED_TRAFFIC_MEDIUM_SPEED: {
        'way/254010814',
        'way/596760204',
        'way/26816187',
        # 'way/152645928', # Lagos testcase commented out for speed reasons
    },
    # https://www.openstreetmap.org/way/254010814 highway=residential and cycleway:both=lane and maxspeed=50
    # https://www.openstreetmap.org/way/596760204 highway=tertiary and cycleway:both=no and maxspeed=50
    # https://www.openstreetmap.org/way/26816187 highway=residential and maxspeed:type=DE:urban
    # https://www.openstreetmap.org/way/152645928 highway=residential (commented out for speed reasons)
    PathCategory.SHARED_WITH_MOTORISED_TRAFFIC_HIGH_SPEED: {'way/372602113', 'way/189054375', 'way/155281272'},
    # https://www.openstreetmap.org/way/372602113 highway=tertiary and mayspeed=70 and cycleway:both=no
    # https://www.openstreetmap.org/way/189054375 highway=unclassified and maxspeed:type=DE:rural
    # https://www.openstreetmap.org/way/155281272 highway=secondary and maxspeed=70 and cycleway:left=no and cyclewayright=separate
    PathCategory.REQUIRES_DISMOUNTING: {
        'way/810025053',
        'way/131478149',
        'way/24968605',
        'way/27797958',
        'way/87956068',
    },
    # https://www.openstreetmap.org/way/810025053 highway=footway and bicycle=dismount
    # https://www.openstreetmap.org/way/131478149 highway=steps and ramp:bicycle=yes
    # https://www.openstreetmap.org/way/24968605 highway=steps and ramp=yes and ramp:stroller=yes
    # https://www.openstreetmap.org/way/27797958 railway=platform
    # https://www.openstreetmap.org/way/87956068 highway=track and ford=yes
    PathCategory.PEDESTRIAN_EXCLUSIVE: {'way/26028197', 'way/870757384', 'way/694458151'},
    # https://www.openstreetmap.org/way/26028197 highway=footway and bicycle=no
    # https://www.openstreetmap.org/way/870757384 highway=pedestrian and bicycle=no
    # https://www.openstreetmap.org/way/694458151 footway=yes
    PathCategory.RESTRICTED_ACCESS: {'way/320034117', 'way/849049867', 'way/25805786'},
    # https://www.openstreetmap.org/way/849049867 highway=service and access=no and bus=yes
    # https://www.openstreetmap.org/way/25805786 highway=service and access=private
    # https://www.openstreetmap.org/way/320034117 highway=footway and access=private
    PathCategory.UNKNOWN: set(),
}


@pytest.fixture(scope='module')
def bpolys():
    """Small bounding boxes."""
    bpolys = gpd.GeoSeries(
        data=[
            # large box around Heidelberg:
            shapely.box(8.65354, 49.37019, 8.7836, 49.4447),
            # Lagos: (small box for way/152645928, commented out for speed reasons)
            # shapely.box(3.354680, 6.444900, 3.369928, 6.455165),
            # Ford:
            shapely.box(8.786665, 49.370061, 8.786965, 49.370305),
            # Motorroad:
            shapely.box(8.491960, 49.476107, 8.496047, 49.477822),
        ],
        crs='EPSG:4326',
    )
    return bpolys


@pytest.fixture(scope='module')
def request_ohsome(bpolys):
    return partial(
        OhsomeClient(
            user_agent='HeiGIT Climate Action Walkability Tester', retry=Retry(total=1)
        ).elements.geometry.post,
        bpolys=bpolys,
        properties='tags',
        time='2024-01-01',
        timeout=60,
    )


@pytest.fixture(scope='module')
def id_filter() -> str:
    """Optimization to make the ohsome API request time faster."""
    full_ids = set().union(*validation_objects.values())
    return f'id:({",".join(full_ids)})'


@pytest.fixture(scope='module')
def osm_return_data(request_ohsome: partial[OhsomeResponse], id_filter: str) -> pd.DataFrame:
    osm_line_data = request_ohsome(filter=f'({ohsome_filter("line")})  and ({id_filter})').as_dataframe(
        multi_index=False
    )
    osm_line_data['category'] = osm_line_data.apply(apply_path_category_filters, axis=1)

    osm_polygon_data = request_ohsome(filter=f'({ohsome_filter("polygon")})  and ({id_filter})').as_dataframe(
        multi_index=False
    )
    osm_polygon_data['category'] = osm_polygon_data.apply(apply_path_category_filters, axis=1)

    return pd.concat([osm_line_data, osm_polygon_data])


@pytest.mark.parametrize('category', validation_objects)
def test_construct_filter_validate(osm_return_data: pd.DataFrame, category: PathCategory):
    osm_return_data = osm_return_data[osm_return_data['category'] == category]

    assert set(osm_return_data['@osmId']) == validation_objects[category]


def test_fetch_osm_data(expected_compute_input, responses_mock):
    with open('resources/test/ohsome_line_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    expected_osm_data = gpd.GeoDataFrame(
        data={
            '@other_tags': [{'bicycle': 'no'}],
        },
        geometry=[shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)])],
        crs=4326,
    )
    computed_osm_data = fetch_osm_data(expected_compute_input.get_aoi_geom(), 'dummy=yes', OhsomeClient())
    geopandas.testing.assert_geodataframe_equal(computed_osm_data, expected_osm_data, check_like=True)


def test_boost_route_members(expected_compute_input, responses_mock):
    """Testing the correct behavior of the boosting."""
    with open('resources/test/ohsome_line_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    paths_input = gpd.GeoDataFrame(
        data={
            'category': [
                PathCategory.NOT_BIKEABLE,  # overlapping, but not in boostable categories -> should stay
                PathCategory.UNKNOWN,  # not overlapping, but in boostable categories -> should stay
                PathCategory.UNKNOWN,  # overlapping and in boostable categories -> should be boosted
            ]
        },
        geometry=[
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3005, 48.2208), (12.3010, 48.22)]),
            shapely.LineString([(12.3, 48.22), (12.3, 48.2205), (12.3005, 48.22)]),
        ],
        crs=4326,
    )

    expected_output = pd.Series(
        data=[PathCategory.NOT_BIKEABLE, PathCategory.UNKNOWN, PathCategory.DESIGNATED_EXCLUSIVE], name='category'
    )

    computed_output = boost_route_members(expected_compute_input.get_aoi_geom(), paths_input, OhsomeClient())
    pd.testing.assert_series_equal(computed_output, expected_output)


def test_boost_route_members_overlapping_routes(expected_compute_input, responses_mock):
    """
    Testing the case if routes are overlapping themselves (i.e. two identical lines in the response of the route query).
    """
    with open('resources/test/ohsome_route_response.geojson', 'rb') as vector:
        responses_mock.post(
            'https://api.ohsome.org/v1/elements/geometry',
            body=vector.read(),
        )

    paths_input = gpd.GeoDataFrame(
        data={'category': [PathCategory.UNKNOWN]},
        geometry=[
            shapely.LineString([(0, 0), (1, 1)]),
        ],
        crs=4326,
    )

    expected_output = pd.Series(data=[PathCategory.DESIGNATED_EXCLUSIVE], name='category')

    computed_output = boost_route_members(expected_compute_input.get_aoi_geom(), paths_input, OhsomeClient())
    pd.testing.assert_series_equal(computed_output, expected_output)


def test_fix_geometry_collection():
    expected_geom = shapely.LineString([(0, 0), (1, 0), (1, 1)])

    geometry_collection_input = shapely.GeometryCollection(
        [
            shapely.Point(-1, -1),
            expected_geom,
        ]
    )
    point_input = shapely.Point(-1, -1)

    input_output_map = {
        'unchanged': {'input': expected_geom, 'output': expected_geom},
        'extracted': {'input': geometry_collection_input, 'output': expected_geom},
        'ignored': {'input': point_input, 'output': shapely.LineString()},
    }
    for _, map in input_output_map.items():
        computed_geom = fix_geometry_collection(map['input'])
        assert_geometries_equal(computed_geom, map['output'])


def test_get_color():
    expected_output = pd.Series([Color('#006837'), Color('#feffbe'), Color('#a50026')])

    expected_input = pd.Series([1.0, 0.5, 0.0])
    computed_output = get_color(expected_input)

    pd.testing.assert_series_equal(computed_output, expected_output)


@pytest.mark.parametrize('geometry_type', ['line', 'polygon'])
def test_ohsome_filter(geometry_type):
    verify(ohsome_filter(geometry_type), options=NamerFactory.with_parameters(geometry_type))
