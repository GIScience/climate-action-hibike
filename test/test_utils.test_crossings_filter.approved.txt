geometry:point and type:node and (crossing=zebra or crossing:markings=zebra or crossing_ref=zebra)
